### PR TITLE
net: lwm2m: Add shell command to enable timeseries cache

### PR DESCRIPTION
**Pull request by SeppoTakalo**
Add shell command
```
  cache   :Enable data cache for resource
           cache PATH NUM
           PATH is LwM2M path
           NUM how many elements to cache
```

### CODE REVIEW

**Review by jheiskan81 — DISMISSED** [review]

File: subsys/net/lib/lwm2m/lwm2m_shell.c [hunk 1]
```diff
@@ -466,6 +470,50 @@ static int cmd_unlock(const struct shell *sh, size_t argc, char **argv)
...
+		shell_error(sh, "Out of memory\n");
+		return -ENOEXEC;
+	}
+	strncpy(path, argv[1], plen - 1);
+	path[plen - 1] = '\0';
+
+	int len = atoi(argv[2]);
+
+	if (len < 1) {
+		shell_error(sh, "Size must be 1 or more (given %d)\n", len);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_shell.c:493]: `path` leaks here.

File: subsys/net/lib/lwm2m/lwm2m_shell.c [hunk 2]
```diff
@@ -466,6 +470,50 @@ static int cmd_unlock(const struct shell *sh, size_t argc, char **argv)
...
+static int cmd_cache(const struct shell *sh, size_t argc, char **argv)
+{
+	if (argc != 3) {
+		shell_error(sh, "wrong parameters\n");
+		return -EINVAL;
+	}
+
+	size_t plen = strlen(argv[1]) + 1;
+	char *path = k_malloc(plen);
+	if (!path) {
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_shell.c:483]: Personally I prefer to keep all declarations at top of the code, it keeps the rest of the code more compact (no need for empty line then).

**Review by rlubos — COMMENTED** [review]
What if the command is called multiple times for the same resource? It seems that we leak previous allocations then.

File: subsys/net/lib/lwm2m/lwm2m_shell.c (hunk 2 above)
Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m/lwm2m_shell.c:493]: Moved to beginning of function.

File: subsys/net/lib/lwm2m/lwm2m_shell.c (hunk 1 above)
Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m/lwm2m_shell.c]: fixed.

**Comment by SeppoTakalo** [comment]
> What if the command is called multiple times for the same resource? It seems that we leak previous allocations then.

True, we leak.
But after all, this is supposed to be a test code, I don't think shell is intended to be used by the end product.

**Comment by rlubos** [comment]
> > What if the command is called multiple times for the same resource? It seems that we leak previous allocations then.
> 
> True, we leak. But after all, this is supposed to be a test code, I don't think shell is intended to be used by the end product.

Ok, although I think we could at least add some protection for calling the command for the second time (for the same resource). It should be pretty straightforward to do by checking if the cache already exists with `lwm2m_cache_entry_get_by_string()`, and leaving early if it does (along with printing the respective error message).

**Review by jheiskan81 — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by SeppoTakalo** [comment]
Closed as the commit is already cherry-picked and reworked in https://github.com/zephyrproject-rtos/zephyr/pull/53372

**Pull request closed without merging**